File: frontend/lib/booking/utils/seat_logic.dart
import '../cubit/booking_state.dart';

class SeatLogic {
  /// Kiểm tra xem việc chọn ghế [candidate] có tạo ra ghế lẻ (orphan) không.
  /// Trả về true nếu VI PHẠM (tạo ra ghế lẻ).
  static bool wouldCreateOrphan(
    Seat candidate,
    List<Seat> allSeats,
    List<Seat> currentSelectedSeats, {
    bool isCoach45 = false,
    bool isCoach28 = false,
  }) {
    List<Seat> group;

    if (isCoach45) {
      // Logic riêng cho xe 45 chỗ (COACH):
      // - 40 ghế đầu (hàng đôi): KHÔNG áp dụng check ghế lẻ.
      // - 5 ghế cuối: KHÔNG được để trống 1 ghế ở giữa (Inner Orphan), cho phép tối đa 1 ghế lẻ ở bìa.
      final sorted = List<Seat>.from(allSeats)..sort((a, b) => a.id.compareTo(b.id)); // Sort theo ID như Frontend
      final index = sorted.indexWhere((s) => s.id == candidate.id);
      
      if (index < 40) return false; // Hàng đôi -> Cho phép chọn tự do
      
      group = sorted.skip(40).take(5).toList(); // Nhóm 5 ghế cuối
    } else if (isCoach28) {
      // Logic riêng cho xe 28 chỗ (COACH):
      // - 24 ghế đầu (hàng đôi): KHÔNG áp dụng check ghế lẻ.
      // - 4 ghế cuối (hàng ngang): KHÔNG được để trống 1 ghế ở giữa.
      final sorted = List<Seat>.from(allSeats)..sort((a, b) => a.id.compareTo(b.id)); // Sort theo ID
      final index = sorted.indexWhere((s) => s.id == candidate.id);

      if (index < 24) return false; // Hàng đôi -> Cho phép chọn tự do
      
      group = sorted.skip(24).take(4).toList(); // Nhóm 4 ghế cuối
    } else {
      // 1. Xác định nhóm ghế của candidate (Cùng hàng hoặc cùng cụm)
      group = _getSeatGroup(candidate, allSeats);
    }

    // Nếu nhóm chỉ có 1 ghế hoặc ghế này độc lập -> Không vi phạm
    if (group.length <= 1) return false;

    // 2. Tạo danh sách các ghế SẼ ĐƯỢC CHỌN (giả lập)
    // 2. Tạo danh sách các ghế SẼ ĐƯỢC CHỌN (giả lập)
    final simulatedSelection = List<Seat>.from(currentSelectedSeats);
    
    // Đảm bảo dùng ID check
    final exists = simulatedSelection.any((s) => s.id == candidate.id);
    if (!exists) {
      simulatedSelection.add(candidate);
    }
    // LƯU Ý: Không return false ở đây nữa, để cho phép check cả những ghế đang có trong list.

    // 3. Kiểm tra trạng thái của cả nhóm ghế xem có generated gap = 1 không
    // Map trạng thái: 0 = Trống (Available), 1 = Đã/Đang chọn (Ticketed/Selected), -1 = Blocked/Sold
    List<int> states = group.map((seat) {
      if (seat.status != 'AVAILABLE') return 1; // Đã bán coi như đã lấp đầy
      if (simulatedSelection.any((s) => s.id == seat.id)) return 1; // Đang chọn
      return 0; // Trống
    }).toList();

    // 4. Quét tìm orphan (ghế trống) và phân loại
    // Luật mới:
    // - KHÔNG được phép có orphan "kẹp giữa" (Inner Hole size 1). -> Chặn tuyệt đối.
    // - ĐƯỢC phép có tối đa 1 orphan "bìa" (Outer Hole size 1). 
    //   -> Tại sao? Vì nhóm 3 ghế (A, B, C), nếu chọn A, B thì còn lại C (Orphan bìa). Phải cho phép.
    //   -> Nhưng nếu chọn B, thì còn lại A (Bìa) và C (Bìa) -> 2 Orphan -> Chặn.
    
    int outerOrphans = 0;
    int currentGapSize = 0;
    bool gapStarted = false;

    for (int i = 0; i < states.length; i++) {
      if (states[i] == 0) {
        currentGapSize++;
        gapStarted = true;
      } else {
        if (gapStarted) {
          // Kết thúc 1 gap
          // Kiểm tra xem gap này là Inner hay Outer (bắt đầu từ đầu dãy?)
          bool isAtStart = (i - currentGapSize == 0);
          
          if (currentGapSize == 1) {
            if (isAtStart) {
              outerOrphans++;
            } else {
              // Gap kẹp giữa (vì bên trái là 1 (do vòng lặp trước), bên phải là 1 (hiện tại))
              return true; // INNER ORPHAN -> CHẶN NGAY
            }
          }
          currentGapSize = 0;
          gapStarted = false;
        }
      }
    }

    // Check gap cuối cùng (nếu dãy kết thúc bằng 0)
    if (gapStarted) {
      if (currentGapSize == 1) {
        outerOrphans++;
      }
    }

    // Nếu có quá 1 ghế lẻ ở bìa (VD: chọn B, hổng A và C) -> Chặn
    if (outerOrphans > 1) return true;

    return false;
  }

  /// Xác định nhóm các ghế liên quan nhau (Row neighbors)
  static List<Seat> _getSeatGroup(Seat target, List<Seat> allSeats) {
    // 1. Phân loại theo cấu trúc xe dựa trên tổng số ghế
    // Logic này phải khớp với cách Frontend render (SelectBusScreen.dart)
    
    // Tách ghế theo tầng và SORT để đảm bảo thứ tự A->B->C (Column Order)
    final sameFloorSeats = allSeats.where((s) => s.floor == target.floor).toList();
    
    // Sort alpha-numeric chuẩn xác: Prefix (Cột) -> Number (Hàng)
    // VD: A1, A2, ..., A10, B1...
    sameFloorSeats.sort((a, b) {
      // 1. Tách Prefix (Chữ) và Suffix (Số)
      final aPrefix = a.seatNumber.replaceAll(RegExp(r'[0-9]'), '');
      final bPrefix = b.seatNumber.replaceAll(RegExp(r'[0-9]'), '');
      
      final aNumStr = a.seatNumber.replaceAll(RegExp(r'[^0-9]'), '');
      final bNumStr = b.seatNumber.replaceAll(RegExp(r'[^0-9]'), '');
      final aNum = int.tryParse(aNumStr) ?? 0;
      final bNum = int.tryParse(bNumStr) ?? 0;

      // 2. So sánh Prefix trước (Gom nhóm cột A, B, C)
      int prefixCompare = aPrefix.compareTo(bPrefix);
      if (prefixCompare != 0) return prefixCompare;

      // 3. Nếu cùng cột -> So sánh số (để xếp đúng hàng 1, 2, 10...)
      return aNum.compareTo(bNum);
    });
    // Sort theo ID hoặc thứ tự xuất hiện (giả định list truyền vào đã sort đúng như Frontend)
    // Nếu chưa chắc chắn, ta nên sort giống repository: floor ASC, seatNumber ASC.
    // Tuy nhiên, ở Cubit state.seats thường đã được load từ Repo và sort rồi.
    
    // --- LAYOUT 34 GIƯỜNG ---
    if (allSeats.length == 34) {
       // Cấu trúc mỗi tầng Frontend: 3 cột [6, 5, 6]
       return _findRowNeighborsByColumns(target, sameFloorSeats, [6, 5, 6]);
    }

    // --- LAYOUT 41 GIƯỜNG ---
    if (allSeats.length == 41) {
       // Logic phức tạp từ Frontend (SelectBusScreen.dart - SeatLayout41Form):
       // Back Row (5 ghế) = 2 ghế cuối tầng trên + 3 ghế cuối tầng dưới.
       
       // 1. Lấy danh sách (giả định đã sort đúng theo thứ tự hiển thị hoặc row-based)
       // KHÔNG DÙNG _sortSeats (vì nó sort theo cột A->B->C, làm vỡ logic Round Robin của Layout này)
       // Ta tin tưởng vào thứ tự của list gốc (hoặc sort cơ bản id/seatNumber nếu cần)
       // Tuy nhiên, SelectBusScreen dùng logic i % 3, tức là list phải có dạng [A1, B1, C1, A2, B2...]
       
       final lowerSeats = allSeats.where((s) => s.floor == 1).toList();
       final upperSeats = allSeats.where((s) => s.floor == 2).toList();
       
       // 2. Xác định các ghế "Moved" (3 ghế cuối tầng dưới dời lên)
       // Logic gốc: movedSeats = lowerSeats.skip(18).take(3)
       // (Lưu ý: lowerSeats có thể > 18. Nếu < 18 thì take(3) sẽ lấy ít hơn, ko sao)
       final movedSeats = lowerSeats.skip(18).take(3).toList();
       
       // 3. Xác định các ghế "Last Row Upper" (2 ghế cuối tầng trên)
       // Logic gốc: lastRowUpperSeats = upperSeats.skip(18).take(2)
       final lastRowUpperSeats = upperSeats.skip(18).take(2).toList();
       
       // 4. Các ghế còn lại (Main Rows - Thân xe)
       final mainLower = lowerSeats.take(18).toList(); // 18 ghế đầu tầng dưới
       final mainUpper = upperSeats.take(18).toList(); // 18 ghế đầu tầng trên
       
       // 5. Build nhóm "Hàng cuối" (Back Row)
       // Thứ tự Visual trong Row: [ ...lastRowUpperSeats, ...movedSeats ]
       final backRow = [...lastRowUpperSeats, ...movedSeats];
       
       // 6. Kiểm tra Target thuộc nhóm nào?
       
       // Check ID trong Back Row
       if (backRow.any((s) => s.id == target.id)) {
         return backRow; // Trả về nguyên nhóm 5 ghế
       }
       
       // Nếu thuộc Main Lower (18 ghế) -> Dùng ROUND ROBIN (3 cột)
       if (mainLower.any((s) => s.id == target.id)) {
         return _findRowNeighborsByRoundRobin(target, mainLower, 3);
       }
       
       // Nếu thuộc Main Upper (18 ghế) -> Dùng ROUND ROBIN (3 cột)
       if (mainUpper.any((s) => s.id == target.id)) {
         return _findRowNeighborsByRoundRobin(target, mainUpper, 3);
       }
       
       // Fallback
       return [target];
    }
    
    // --- LAYOUT 44/45 GIƯỜNG (Mặc định) ---
    // Thường dùng Round Robin
    return _findRowNeighborsByRoundRobin(target, sameFloorSeats, 3);
  }

  // Logic tìm hàng xóm cho Layout chia block (VD: [6, 5, 6])
  static List<Seat> _findRowNeighborsByColumns(Seat target, List<Seat> floorSeats, List<int> colCounts) {
    // 1. Xây dựng cấu trúc Columns
    List<List<Seat>> columns = [];
    int index = 0;
    for (int count in colCounts) {
      if (index + count <= floorSeats.length) {
        columns.add(floorSeats.sublist(index, index + count));
        index += count;
      } else {
        // Hết ghế
        columns.add(floorSeats.sublist(index));
        index = floorSeats.length;
      }
    }

    // 2. Tìm vị trí (colIndex, rowIndex) của target
    int targetColIndex = -1;
    int targetRowIndex = -1;

    for (int c = 0; c < columns.length; c++) {
      // Dùng ID để tìm kiếm thay vì object reference (tránh lỗi instance khác nhau)
      int r = columns[c].indexWhere((s) => s.id == target.id);
      if (r != -1) {
        targetColIndex = c;
        targetRowIndex = r;
        break;
      }
    }

    if (targetRowIndex == -1) return [target];

    // 3. Lấy các ghế ở cùng rowIndex từ các cột khác
    List<Seat> neighbors = [];
    for (int c = 0; c < columns.length; c++) {
      if (targetRowIndex < columns[c].length) {
        neighbors.add(columns[c][targetRowIndex]);
      }
    }
    
    // Sort lại cho đúng trật tự trái -> phải (để logic check gap hoạt động đúng)
    // Giả định thứ tự cột từ 0..N là trái -> phải
    return neighbors;
  }

  // Logic tìm hàng xóm cho Layout chia Round Robin (i % cols)
  static List<Seat> _findRowNeighborsByRoundRobin(Seat target, List<Seat> floorSeats, int colCount) {
     int index = floorSeats.indexOf(target);
     if (index == -1) return [target];

     // Row index trong Round Robin chính là: index div colCount
     int rowIndex = index ~/ colCount;
     
     // Tìm tất cả ghế có cùng rowIndex
     List<Seat> neighbors = [];
     for (int i = 0; i < floorSeats.length; i++) {
       if ((i ~/ colCount) == rowIndex) {
         neighbors.add(floorSeats[i]);
       }
     }
     
     return neighbors;
  }
  /// Tìm các ghế không hợp lệ trong danh sách đã chọn (dùng cho Auto-Deselect)
  static List<Seat> findInvalidSeats(List<Seat> allSeats, List<Seat> currentSelected, {bool isCoach45 = false, bool isCoach28 = false}) {
    final invalid = <Seat>{};
    for (final seat in currentSelected) {
      if (wouldCreateOrphan(seat, allSeats, currentSelected, isCoach45: isCoach45, isCoach28: isCoach28)) {
        invalid.add(seat);
      }
    }
    return invalid.toList();
  }
  
  static void _sortSeats(List<Seat> seats) {
    seats.sort((a, b) {
      final aPrefix = a.seatNumber.replaceAll(RegExp(r'[0-9]'), '');
      final bPrefix = b.seatNumber.replaceAll(RegExp(r'[0-9]'), '');
      
      final aNumStr = a.seatNumber.replaceAll(RegExp(r'[^0-9]'), '');
      final bNumStr = b.seatNumber.replaceAll(RegExp(r'[^0-9]'), '');
      final aNum = int.tryParse(aNumStr) ?? 0;
      final bNum = int.tryParse(bNumStr) ?? 0;

      int prefixCompare = aPrefix.compareTo(bPrefix);
      if (prefixCompare != 0) return prefixCompare;

      return aNum.compareTo(bNum);
    });
  }
}

File: web/src/lib/booking/seatLogic.ts
import { Seat } from "@/types/seat";

export class SeatLogic {
    /**
     * Sorting helper to match Dart/Flutter logic:
     * Sorts seats by Prefix (A, B, C...) then by Number (1, 2, 10...).
     */
    static sortSeats(seats: Seat[]): Seat[] {
        return [...seats].sort((a, b) => {
            // Extract prefix (letters)
            const aPrefix = a.seatNumber.replace(/[0-9]/g, '');
            const bPrefix = b.seatNumber.replace(/[0-9]/g, '');

            // Extract number
            const aNumStr = a.seatNumber.replace(/[^0-9]/g, '');
            const bNumStr = b.seatNumber.replace(/[^0-9]/g, '');
            const aNum = parseInt(aNumStr) || 0;
            const bNum = parseInt(bNumStr) || 0;

            const prefixCompare = aPrefix.localeCompare(bPrefix);
            if (prefixCompare !== 0) return prefixCompare;

            return aNum - bNum;
        });
    }

    /**
     * Checks if selecting [candidate] would create an orphan seat violation.
     * Returns TRUE if violation (should block selection).
     */
    static wouldCreateOrphan(
        candidate: Seat,
        allSeats: Seat[],
        currentSelectedSeats: Seat[],
        totalSeats = 0,
        seatType = '',
        othersSelecting: Record<number, { userId: string }> = {},
        currentUserId = ''
    ): boolean {
        // Find group
        const group = this.getSeatGroup(candidate, allSeats, totalSeats, seatType);

        // Single seat or empty group -> No violation possible
        if (group.length <= 1) return false;

        // Base states (0 = Available, 1 = Taken by others (DB or SSE Lock))
        const baseStates = group.map(seat => {
            if (!seat.isAvailable) return 1;
            if (othersSelecting[seat.id] && othersSelecting[seat.id].userId !== currentUserId) return 1;
            return 0;
        });

        // Simulated states (0 = Available, 1 = Taken by others OR Selected by ME)
        const simulatedSelection = [...currentSelectedSeats];
        if (!simulatedSelection.some(s => s.id === candidate.id)) {
            simulatedSelection.push(candidate);
        }

        const simStates = group.map(seat => {
            if (!seat.isAvailable) return 1;
            if (othersSelecting[seat.id] && othersSelecting[seat.id].userId !== currentUserId) return 1;
            if (simulatedSelection.some(s => s.id === seat.id)) return 1;
            return 0;
        });

        // Find all '0' blocks (contiguous available seats)
        const simBlocks = this.getZeroBlocks(simStates);
        const baseBlocks = this.getZeroBlocks(baseStates);

        /*
            Orphan Rules:
            - Rule 1 (Edge-Packing): You cannot pick a seat that fragments an existing empty block into smaller parts.
              (i.e. number of empty blocks cannot increase). This forces picking from the edges of a row/mattress.
            - Rule 2 (No Single Orphan): Any gap of exactly 1 empty seat is BLOCKED.
              EXCEPTION: If the original available block had size <= 2, leaving 1 is allowed. 
              (Because it's impossible to book 1 seat from a pair without leaving 1).
        */
        if (simBlocks.length > baseBlocks.length) {
            return true; // Rule 1 Violation
        }

        // Rule 2: No single Inner Orphan (sandwiched gap).
        // Outer orphans (size 1 gap touching the physical boundary) are ALLOWED.
        for (const simBlock of simBlocks) {
            if (simBlock.length === 1) {
                const idx = simBlock[0];
                // Check if it's strictly an INNER gap (not touching 0 or length - 1)
                if (idx > 0 && idx < group.length - 1) {
                    const baseBlock = baseBlocks.find(b => b.includes(idx));
                    // If it was already an inner orphan in base state, we are immune
                    if (baseBlock && baseBlock.length > 1) {
                        return true; // Invalid inner orphan created
                    }
                }
            }
        }

        return false;
    }

    private static getZeroBlocks(states: number[]): number[][] {
        const blocks: number[][] = [];
        let currentBlock: number[] = [];
        for (let i = 0; i < states.length; i++) {
            if (states[i] === 0) {
                currentBlock.push(i);
            } else {
                if (currentBlock.length > 0) {
                    blocks.push(currentBlock);
                    currentBlock = [];
                }
            }
        }
        if (currentBlock.length > 0) blocks.push(currentBlock);
        return blocks;
    }

    /**
     * Identifies the row/group of neighbors for a target seat.
     */
    private static getSeatGroup(target: Seat, allSeats: Seat[], totalSeats = 0, seatType = ''): Seat[] {
        const allLen = totalSeats > 0 ? totalSeats : allSeats.length;

        // Layout 45 (Coach)
        if (allLen === 45) {
            const sortedAll = [...allSeats].sort((a, b) => a.id - b.id);
            const index = sortedAll.findIndex(s => s.id === target.id);
            // First 40 seats (10 rows of 4). Group them into pairs of 2.
            if (index < 40) {
                const pairStart = Math.floor(index / 2) * 2;
                return sortedAll.slice(pairStart, pairStart + 2);
            }
            // Last 5 seats -> Orphan Check Group
            return sortedAll.slice(40);
        }

        // Layout 28/29 (Limousine)
        if (allLen === 28 || allLen === 29) {
            const sortedAll = [...allSeats].sort((a, b) => a.id - b.id);
            const index = sortedAll.findIndex(s => s.id === target.id);
            // First 24 seats (6 rows of 4). Group them into pairs of 2.
            if (index < 24) {
                const pairStart = Math.floor(index / 2) * 2;
                return sortedAll.slice(pairStart, pairStart + 2);
            }
            // Last 4 or 5 seats -> Orphan Check Group
            return sortedAll.slice(24);
        }

        // 1. Group by floor
        const floorSeats = allSeats.filter(s => s.floor === target.floor);
        // Sort by Column logic A1, A2...
        const sortedFloorSeats = this.sortSeats(floorSeats);

        // Layout 34 (3 cols: 6-5-6 heuristic from Dart)
        if (allLen === 34) {
            return this.findRowNeighborsByColumns(target, sortedFloorSeats, [6, 5, 6]);
        }

        // Layout 41 (Special logic)
        if ((allLen >= 35 && allLen <= 44 && (seatType === 'SLEEPER' || seatType === 'LIMOUSINE')) || allLen === 41) {
            // Logic port from Dart:
            // Back Row (5 seats) = 2 last upper + 3 last lower
            const sortedAll = [...allSeats].sort((a, b) => a.id - b.id);
            // Ensure we use the sorted IDs for consistent logic with Layout
            const lowerS = sortedAll.filter(s => s.floor === 1);
            const upperS = sortedAll.filter(s => s.floor === 2);

            const movedSeats = lowerS.slice(18, 21); // take 3 skipping 18
            const lastRowUpper = upperS.slice(18, 20); // take 2 skipping 18

            const mainLower = lowerS.slice(0, 18);
            const mainUpper = upperS.slice(0, 18);

            const backRow = [...lastRowUpper, ...movedSeats].sort((a, b) => a.id - b.id);

            if (backRow.some(s => s.id === target.id)) return backRow;
            if (mainLower.some(s => s.id === target.id)) return this.findRowNeighborsByRoundRobin(target, mainLower, 3);
            if (mainUpper.some(s => s.id === target.id)) return this.findRowNeighborsByRoundRobin(target, mainUpper, 3);

            return [target];
        }

        // Default Round Robin (3 cols)
        return this.findRowNeighborsByRoundRobin(target, sortedFloorSeats, 3);
    }

    private static findRowNeighborsByColumns(target: Seat, floorSeats: Seat[], colCounts: number[]): Seat[] {
        const columns: Seat[][] = [];
        let index = 0;

        for (const count of colCounts) {
            if (index + count <= floorSeats.length) {
                columns.push(floorSeats.slice(index, index + count));
                index += count;
            } else {
                columns.push(floorSeats.slice(index));
                index = floorSeats.length;
            }
        }

        let targetColIndex = -1;
        let targetRowIndex = -1;

        for (let c = 0; c < columns.length; c++) {
            const r = columns[c].findIndex(s => s.id === target.id);
            if (r !== -1) {
                targetColIndex = c;
                targetRowIndex = r;
                break;
            }
        }

        if (targetRowIndex === -1) return [target];

        const neighbors: Seat[] = [];
        for (let c = 0; c < columns.length; c++) {
            if (targetRowIndex < columns[c].length) {
                neighbors.push(columns[c][targetRowIndex]);
            }
        }
        return neighbors;
    }

    private static findRowNeighborsByRoundRobin(target: Seat, floorSeats: Seat[], colCount: number): Seat[] {
        const index = floorSeats.findIndex(s => s.id === target.id);
        if (index === -1) return [target];

        const rowIndex = Math.floor(index / colCount);

        return floorSeats.filter((_, i) => Math.floor(i / colCount) === rowIndex);
    }

    /**
     * Finds seats that become invalid (create orphans) in the current selection
     * due to the removal of other seats or existing state.
     * Ported from Flutter `SeatLogic.findInvalidSeats`.
     */
    static findInvalidSeats(
        allSeats: Seat[],
        currentSelected: Seat[],
        totalSeats = 0,
        seatType = '',
        othersSelecting: Record<number, { userId: string }> = {},
        currentUserId = ''
    ): Seat[] {
        const invalid: Seat[] = [];
        for (const seat of currentSelected) {
            // Check if THIS seat causes a violation in the CURRENT set
            if (this.wouldCreateOrphan(seat, allSeats, currentSelected, totalSeats, seatType, othersSelecting, currentUserId)) {
                invalid.push(seat);
            }
        }
        return invalid;
    }
}
